### DISCUSSION TITLE
Binarize and Grayscale Image

### DISCUSSION THREAD

**Discussion by mirfani28** [teamalgoritma/community#498] [Q&A] [answered]
Hi Teman-Teman dan Instruktur Algo

Saya sedang coba latihan neural network menggunakan dataset MNIST sign language

Saya notice ketika glimpse sedikit ke dataset, ada satu dan beberapa hal yang menarik :

![image](https://user-images.githubusercontent.com/90766716/209080015-1b4d6c10-6cf1-4c67-9e73-35cd47f1bd38.png)
Ini adalah data file gambar asli yang diambil dari contoh dataset

![image](https://user-images.githubusercontent.com/90766716/209082129-a86514b1-005a-467b-9d92-1e35bbca6e31.png)
Ini adalah gambar yang di generate dari data pixel .csv

![image](https://user-images.githubusercontent.com/90766716/209081929-3bb49fa6-5370-4222-a4aa-ea1b3a90db0f.png)
Ini adalah gambar yang di generate dari data pixel .csv MNIST digit

Karena saya belum menemukan sumber yang mentransform data asli (.jpeg .png) menjadi data .csv, saya tidak tahu apakah **proses grayscaling** kedua dataset tersebut ditransform dengan parameter yang sama atau tidak.

Setau saya, proses grayscaling pun ada perbedaan dari seberapa kuat channel tiap warna difilter
Seperti di gambar ini :
![image](https://user-images.githubusercontent.com/90766716/209082614-9bcfa634-16a6-4ce0-a41a-d14843591542.png)

Pertanyaan Saya :
1. Saya membayangkan apabila misalnya data MNIST sign language diambil dari foto yang berbeda beda (misal di ruangan yg agak sedikit terang dan agak sedikit gelap), nilai informasi yang hilang pun akan berbeda. Lantas berarti model yang dihasilkan hanya dapat memprediksi nilai secara akurat hanya pada data foto yang diambil dari kondisi ruangan yang sama. Betul kan?

2. Kapan kita harus mempertimbangkan menggunakan grayscale? Sedangkan kita tau nilai informasi yang hilang di setiap foto akan berbeda-beda. 

3. Adakah proses grayscale yang **adaptif** pada setiap gambar? yang misal menjaga agar gambar yang digrayscale tidak menjadi terlalu gelap atau terlalu putih sehingga nilai informasi yang hilang sedikit.

![image](https://user-images.githubusercontent.com/90766716/209085011-0dff8c81-5ce0-44d1-94c4-eaf9664e59f5.png)
4.  Apabila tujuannya untuk mempercepat komputasi, kenapa hanya scaling nilai 1-255 jadi range 0.0-1.0 ? kenapa tidak dibinarize sekalian jadi hanya ada nilai 0 dan 1 seperti gambar di atas? (Untuk kasus MNIST digit)

5. Apabila misal data MNIST digit dibinarize jadi hanya ada nilai 0 dan 1 pada pixel, masih relevan kah menggunakan Neural Network untuk prediksi jenis kelas? 

6. Untuk proses latihan MNIST digit, apakah proses menjadikan data gambar menjadi sebuah data frame dengan kolom pixel 1-784 adalah proses flattening? Apakah tujuannya agar bisa langsung diproses menggunakan Neural Network Fully Connected Layer?

7. Apa step by step yang umum ketika kita mempersiapkan dan memeriksa dataset untuk kasus klasifikasi gambar? Kapan kita memutuskan untuk menggunakan metode Convolutional NN atau metode grayscaling dan flattening seperti di dataset MNIST?

Thanks

**Comment by yosiaazarya** [comment]
Hi Bapak, team kami sedang mencarikan referensi serta jawaban untuk mendukung diskusi yang Bapak berikan, mohon ditunggu yaa

**Comment by dwiknrd** [comment, marked as answer]
Hai [mirfani28](https://github.com/mirfani28) terima kasih atas pertanyaannya, disini saya akan coba membantu menjawab beberapa pertanyaan diatas, semoga bisa bermanfaat.

1. Pada dasarnya, menggunakan model neural network basic untuk data gambar akan menggunakan semua fitur pada data, dan semua fitur akan dianggap penting. Dalam proses belajarnya, neural network juga akan melakukan fitur selection sendiri yang dapat dibilang blackbox, kita tidak dapat simpulkan bahwa akurat hanya pada data foto yang diambil dari kondisi ruangan yang sama. 
2. Pertimbangan menggunakan warna grayscale adalah untuk memperkecil fitur yang digunakan sehingga waktu komputasi akan semakin cepat. Pada praktiknya kita dapat mengubah data kita menjadi bentuk grayscale apabila pada model tidak dibutuhkan adanya perbedaan warna melainkan hanya bentuk. Contoh kasus yang membutuhkan adanya warna adalah ketika membuat model untuk membedakan warna traffic light, dll. Sebaliknya, apabila anda hanya membutuhkan model yang dapat mendeteksi bentuk atau pola pada gambar tanpa memerhatikan warna seperti mnist digit dan mnist sign anda dapat melakukan preprocessing data menjadi bentuk grayscale. Sebagai tambahan referensi terkait penggunaan data RGB dan grayscale pada image preprocessing: https://www.researchgate.net/post/Which-is-best-for-image-classification-RGB-or-grayscale
3. Untuk mengubah gambar menjadi bentuk grayscale pada fungsi yang ada pada keras hanya terdapat 1 metode. Apabila anda ingin membuat model berdasarkan pola gambar yang ada pada data anda dapat menggunakan arsitektur deep learning lain yang cocok untuk data gambar yaitu Convolutional Neural Network (CNN). 
4. Benar, apabila gambar diubah menjadi binary image akan mempercepat komputasi, namun untuk mengubah gambar menjadi binary image membutuhkan teknik tertentu, berikut adalah dokumentasi panduan seputar threshold: https://homepages.inf.ed.ac.uk/rbf/CVonline/LOCAL_COPIES/MORSE/threshold.pdf
5. Pada dasarnya, apabila kita mengubah data gambar menjadi biner yaitu 0 dan 1 maka apabila kita memiliki katakanlah bobot besar yang dikalikan dengan data 0 maka akan menghasilkan nilai 0 artinya sangat tidak berharga, berlaku sebaliknya apabila dikali dengan 1 akan sangat berharga yang ditandai dengan nilai yang besar.
6. Betul, proses tersebut adalah proses yang disebut flattening atau mengubah multi dimesi menjadi single dimensi supaya dapat terkoneksi secara sekuensial dengan layer berikutnya.
7. Workflow secara umum melakukan image classification seperti pada dokumentasi berikut: https://keras.io/examples/vision/mnist_convnet/. Prinsip dari tahapan preprocessing image sebelum dilakukan proses training adalah supaya data yang digunakan dapat di fitting pada arsitektur yang sudah dibuat.
Penggunaan metode yang digunakan pada eksperimen machine learning merupakan tahapan yang tidak pasti. Beberapa metode perlu kita gunakan tergantung dengan kondisi data yang dimiliki. Matode flattening akan diperlukan ketika data yang kita olah berupa gambar 2 dimensi dan akan dijadikan data training pada arsitektur neural network.
Penggunaan arsitektur CNN menurut penilitian memiliki hasil performa lebih bagus dibandingkan menggunakan arsitektur neural network basic. Karena pada CNN akan menggunakan layer convolutional yang dapat mengekstrak informasi penting pada gambar menggunakan teknik filter yang ada pada layer tersebut.

Semoga jawaban tersebut dapat memberikan insight.

Regards,
Dwi

**Reply by mirfani28** [reply]
thanks mbak, ini banyak menjawab dari hasil latihan saya untuk LBB.
kebetulan saya pake dua metode tsb untuk modelling machine learningnya..
thanks 😄
